### PR TITLE
Various utils used in Morum

### PR DESCRIPTION
**Pull request by Harrm**

### CODE REVIEW

**Comment by turuslan** [comment]
[Suggestions branch](https://github.com/qdrvm/qtils/tree/suggest/harrm):
- reverted "include" to "src"
- renamed `BytesInN`, `BytesOutN`
- moved `BitSpan`, `FixedByteVector`, `array_from_span`, `array_from_span_fill` to separate headers.
- removed `OptRef` formatter (ambigous, why not format as "None" and "Some({})")
- moved `QTILS_UNIQUE` macro to header
- fixed `QTILS_UNWRAP` macro variable name
- moved `print` and `forward_like` to "cxx23"

**Comment by Harrm** [comment]
>     * reverted "include" to "src"
Why? It's a directory with public header files. 

>     * renamed `BytesInN`, `BytesOutN`
Honestly I don't like these names and find them pretty confusing in some cases. I think they are fine in other cases, for example as function arguments, so it's okay to keep both aliases and use them as seems contextually more readable.

>     * moved `BitSpan`, `FixedByteVector`, `array_from_span`, `array_from_span_fill` to separate headers.
Okay, sure.

>     * removed `OptRef` formatter (ambigous, why not format as "None" and "Some({})")
It's pretty convenient to have it though, maybe discuss tomorrow on status meeting what it should print?

>     * moved `QTILS_UNIQUE` macro to header
Okay.

>     * fixed `QTILS_UNWRAP` macro variable name
Okay.

>     * moved `print` and `forward_like` to "cxx23"

Also, regarding outcome - outcome is the name of the library which has a result class template, why name the alias Outcome?

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+build*
+cmake-build*
+
+.vscode
+.cache
```
Comment (xDimon) [inline .gitignore:5]:
```suggestion
.cache
.idea
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -24,7 +24,13 @@ set(CMAKE_CXX_EXTENSIONS OFF)
...
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
-include(GNUInstallDirs)
+option(QTILS_BUILD_TESTS "Build tests" OFF)
+
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+    option(QTILS_ENABLE_ASSERT "Enable assertions" ON)
+else()
+    option(QTILS_ENABLE_ASSERT "Enable assertions" OFF)
+endif()
```
Comment (turuslan) [inline CMakeLists.txt:33]:
```suggestion
if (CMAKE_BUILD_TYPE STREQUAL "Debug")
    set(QTILS_ENABLE_ASSERT_DEFAULT ON)
else()
    set(QTILS_ENABLE_ASSERT_DEFAULT OFF)
endif()
option(QTILS_ENABLE_ASSERT "Enable assertions" ${QTILS_ENABLE_ASSERT_DEFAULT})
```

File: test/qtils_compiles.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// IWYU pragma: begin_keep
+#include <qtils/append.hpp>
```
Comment (turuslan) [inline test/qtils_compiles.cpp:2]: Need cmake or script to generate/update this file

File: src/qtils/bytes.hpp [hunk 4]
```diff
@@ -11,17 +11,146 @@
...
+
+  template <class FmtContext>
+  FmtContext::iterator format(qtils::ByteSpan s, FmtContext &ctx) const {
+    auto out = ctx.out();
+    for (auto i : s.subspan(0, s.size() - 1)) {
+      std::format_to(out, "{:02x}", i);
+    }
+    if (!s.empty()) {
+      std::format_to(out, "{:02x}", s.back());
+    }
```
Comment (xDimon) [inline src/qtils/bytes.hpp]: How about to print `<empty>` for 0-size span?

File: src/qtils/bytes.hpp [hunk 5]
```diff
@@ -11,17 +11,146 @@
...
+
+  template <class FmtContext>
+  FmtContext::iterator format(const qtils::Bytes &v, FmtContext &ctx) const {
+    auto out = ctx.out();
+    for (auto i : qtils::ByteSpan{v}.subspan(0, v.size() - 1)) {
+      std::format_to(out, "{:02x}", i);
+    }
+    if (!v.empty()) {
+      std::format_to(out, "{:02x}", v.back());
+    }
```
Comment (xDimon) [inline src/qtils/bytes.hpp]: And here

File: src/qtils/optional_ref.hpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  template <class FmtContext>
+  FmtContext::iterator format(
+      qtils::OptionalRef<T> opt, FmtContext &ctx) const {
+    auto out = ctx.out();
+    if (opt) {
+      std::format_to(out, "{}", *opt);
+    } else {
+      std::format_to(out, "nullopt");
```
Comment (xDimon) [inline src/qtils/optional_ref.hpp]:
```suggestion
      std::format_to(out, "<none>");
```

File: src/qtils/optional_ref.hpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    }
+    return out;
+  }
+};
```
Comment (xDimon) [inline src/qtils/optional_ref.hpp]: Add `InsertNewlineAtEOF: true` line to **.clang-format** file to insert EOL at EOF automatic

File: src/qtils/outcome.hpp
Comment (xDimon) [inline src/qtils/outcome.hpp:1]: How about rename this file to **result.hpp**?

File: src/qtils/bytes.hpp (hunk 4 above)
Comment (turuslan) [reply src/qtils/bytes.hpp]: By default bytes are formatted using <qtils/hex.hpp>,
and `"<empty>"` will break hex.

User can enable prefix `"{:0x}"`, quote `"[{}]"`, see words around `"foo {} bar"`, or add length `("{} {}", x.size(), x)`.

Or make separate wrapper like `qtils::Hex` with custom formatter.

File: src/qtils/bytes.hpp [hunk 8]
```diff
@@ -11,17 +11,146 @@
...
+
+    ByteSpan span() const {
+      return ByteSpan{data.data(), size};
+    }
+  };
+
+  inline std::string_view to_string_view(ByteSpan span) {
+    return std::string_view{
+        reinterpret_cast<const char *>(span.data()), span.size()};
+  }
```
Comment (turuslan) [inline src/qtils/bytes.hpp]: `qtils::byte2str`
```suggestion
```

File: src/qtils/bytes.hpp [hunk 9]
```diff
@@ -11,17 +11,146 @@
...
+
+  inline std::string hex(ByteSpan span) {
+    std::string s(span.size() * 2, '\0');
+    for (size_t i = 0; i < span.size(); i++) {
+      std::to_chars(s.data() + 2 * i, s.data() + 2 * i + 1, span[i] >> 4, 16);
+      std::to_chars(
+          s.data() + 2 * i + 1, s.data() + 2 * i + 2, span[i] & 0xF, 16);
+    }
+    return s;
+  }
```
Comment (turuslan) [inline src/qtils/bytes.hpp]: `hex` may cause unnecessary allocations,
formatting is preferred `fmt::format("{:x}", bytes)`.
<qtils/hex.hpp> should be included in ".cpp" where it's used (in ".hpp" too if unavoidable)
```suggestion
```

File: src/qtils/bytes.hpp [hunk 10]
```diff
@@ -11,17 +11,146 @@
...
+    QTILS_ASSERT_GREATER_EQ(span.size(), N);
+    BytesN<N> array;
+    std::ranges::copy_n(span.begin(), N, array.begin());
+    return array;
+  }
+
+  template <size_t N>
+  BytesN<N> array_from_span_fill(ByteSpan span, uint8_t filler = 0) {
+    BytesN<N> array;
+    const auto s = std::min(N, span.size());
```
Comment (turuslan) [inline src/qtils/bytes.hpp]: Ambigous, may truncate instead of fill, rename or add assert

File: src/qtils/bytes.hpp [hunk 11]
```diff
@@ -11,17 +11,146 @@
...
+    std::ranges::copy_n(span.begin(), s, array.begin());
+    std::ranges::fill(
+        std::ranges::subrange(array.begin() + s, array.end()), filler);
+    return array;
+  }
 }  // namespace qtils
+
+template <>
```
Comment (turuslan) [inline src/qtils/bytes.hpp]: Remove formatters,
they are in <qtils/hex.hpp>

File: src/qtils/bytes.hpp [hunk 12]
```diff
@@ -11,17 +11,146 @@
...
     return {v.begin(), v.end()};
   }
+
+  /**
+   * Fixed-capacity vector meant to be allocated on stack
+   * Used to avoid heap allocations for small vectors with known capacity
+   * @tparam N - maximum capacity
+   */
+  template <size_t N>
+  struct FixedByteVector {
```
Comment (turuslan) [inline src/qtils/bytes.hpp]: Move to separate header ("bytes_small.hpp"?).

<qtils/bytes.hpp> is expected to be included for forward declarations,
but this type is used less that simple `Bytes`.

Moving will also remove <qtils/assert.hpp> dependency

Comment (turuslan) [inline src/qtils/bytes.hpp]: This type looks incomplete,
it's missing `size`/`push`/`append`/... to work with it.

Should it be added to qtils later when it's interface will be decided?

File: src/qtils/bytes.hpp [hunk 13]
```diff
@@ -11,17 +11,146 @@
...
+    for (size_t i = 0; i < span.size(); i++) {
+      std::to_chars(s.data() + 2 * i, s.data() + 2 * i + 1, span[i] >> 4, 16);
+      std::to_chars(
+          s.data() + 2 * i + 1, s.data() + 2 * i + 2, span[i] & 0xF, 16);
+    }
+    return s;
+  }
+
+  template <size_t N>
+  BytesN<N> array_from_span(ByteSpan span) {
```
Comment (turuslan) [inline src/qtils/bytes.hpp]: Move `array_from_span` and `array_from_span_fill` to separate header ("array_from_span.hpp"?)

File: src/qtils/macro.hpp [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+#pragma once
+
+#define IDENTITY(a) a
+
+#define CONCAT__(a, b) a##b
+#define CONCAT_(a, b) CONCAT__(a, b)
+#define CONCAT(a, b) CONCAT_(IDENTITY(a), IDENTITY(b))
+
+#define QTILS_UNIQUE_NAME(prefix) CONCAT(prefix, __COUNTER__)
```
Comment (turuslan) [inline src/qtils/macro.hpp]: Duplicates `OUTCOME_UNIQUE` from <qtils/outcome.hpp>

May move macros to separate files ("unique name" and "unexpected" are not related)

```suggestion
#define _QTILS_UNIQUE_NAME_2(x, y) x##y
#define _QTILS_UNIQUE_NAME(prefix, counter) _QTILS_UNIQUE_NAME_2(prefix, counter)
#define QTILS_UNIQUE_NAME_PREFIX(prefix) _QTILS_UNIQUE_NAME(prefix, __COUNTER__)
#define QTILS_UNIQUE_NAME QTILS_UNIQUE_NAME_PREFIX(qtils_unique_name_)
```

File: src/qtils/outcome.hpp
Comment (Harrm) [reply src/qtils/outcome.hpp:1]: It's a wrapper for boost outcome so I think outcome is fitting.

File: CMakeLists.txt (hunk 2 above)
Comment (Harrm) [reply CMakeLists.txt:33]: Is there any functional difference?

File: src/qtils/bytes.hpp (hunk 12 above)
Comment (Harrm) [reply src/qtils/bytes.hpp]: It contains what was required of it. I see no harm in adding it now and adding some new stuff later if it's needed.

File: src/qtils/macro.hpp (hunk 14 above)
Comment (Harrm) [reply src/qtils/macro.hpp]: Ah, I thought I saw it somewhere else.

File: src/qtils/bytes.hpp (hunk 10 above)
Comment (Harrm) [reply src/qtils/bytes.hpp]: I can't come up with a better name, I will add a comment.

File: CMakeLists.txt (hunk 2 above)
Comment (turuslan) [reply CMakeLists.txt:33]: Doesn't duplicate `option`, option name, option comment

File: src/qtils/bytes.hpp (hunk 13 above)
Comment (Harrm) [reply src/qtils/bytes.hpp:81]: Why?

File: src/qtils/byte_utils.hpp [hunk 15]
```diff
@@ -42,4 +42,32 @@ namespace qtils {
   inline BytesOut str2byte(std::span<char> s) {
```
Comment (turuslan) [inline src/qtils/byte_utils.hpp:42]: Revert.
Keep `asVec` in "bytes.hpp" or move to "as_vec.hpp".
Keep "bytestr.hpp".
Add `array_from_span` to new header.

File: src/qtils/macro/common.hpp [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+#pragma once
+
+#define IDENTITY(a) a
+
+#define CONCAT__(a, b) a##b
+#define CONCAT_(a, b) CONCAT__(a, b)
+#define CONCAT(a, b) CONCAT_(IDENTITY(a), IDENTITY(b))
+
+#define QTILS_UNIQUE_NAME(prefix) CONCAT(prefix, __COUNTER__)
```
Comment (turuslan) [inline src/qtils/macro/common.hpp:15]: Are `IDENTITY` and `CONCAT` used somewhere else?
```suggestion
#define _QTILS_UNIQUE_NAME_2(x, y) x##y
#define _QTILS_UNIQUE_NAME(prefix, counter) _QTILS_UNIQUE_NAME_2(prefix, counter)
#define QTILS_UNIQUE_NAME_PREFIX(prefix) _QTILS_UNIQUE_NAME(prefix, __COUNTER__)
#define QTILS_UNIQUE_NAME QTILS_UNIQUE_NAME_PREFIX(qtils_unique_name_)
```

File: src/qtils/macro/unwrap.hpp [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+#define QTILS_UNWRAP(decl, expr) QTILS_UNWRAP_3(QTILS_UNWRAP_NAME, decl, expr)
+
+#define QTILS_UNWRAP_void_2(unwrap_res, expr)   \
+  auto unwrap_res = (expr);                     \
+  if (!unwrap_res) {                            \
+    return std::unexpected(unwrap_res.error()); \
+  }
+
+#define QTILS_UNWRAP_void(expr) QTILS_UNWRAP_void_2(QTILS_UNWRAP_NAME, expr)
```
Comment (turuslan) [inline src/qtils/macro/unwrap.hpp]: See "outcome.hpp".
```suggestion
#define _QTILS_UNWRAP_void(tmp, expr)    \
  auto &&tmp = (expr);                   \
  if (tmp.has_error()) {                 \
    return std::unexpected(tmp.error()); \
  }
#define _QTILS_UNWRAP_out(tmp, out, expr)     \
  _OUTCOME_TRY_void(QTILS_UNWRAP_NAME, expr); \
  out = std::move(tmp).value()

#define QTILS_UNWRAP_void(expr) _QTILS_UNWRAP_void(QTILS_UNWRAP_NAME, expr)
#define QTILS_UNWRAP(out, expr) _QTILS_UNWRAP_out(QTILS_UNWRAP_NAME, out, expr)
```

File: src/qtils/byte_utils.hpp (hunk 15 above)
Comment (Harrm) [reply src/qtils/byte_utils.hpp:42]: I don't want to put every function in a separate file, I find it inconvenient.

File: src/qtils/macro/common.hpp (hunk 16 above)
Comment (Harrm) [reply src/qtils/macro/common.hpp:15]: Not yet, but they are very commonly used macros so I think it's better to keep them separate. Maybe add a prefix though.

**Review by turuslan — APPROVED** [review]

**Review by xDimon — APPROVED** [review]

**Pull request merged**